### ISSUE TITLE
Press Button Proposal

### ISSUE THREAD

**Issue by lukewarlow** [openui/open-ui#1058]
# Press Buttons

## Problem

It's common in UI design to have buttons which toggle a binary state.  These are implemented in varying ways across the web, some involve JS (aria-pressed, or changing the button label), some are semantically questionable (checkbox, switch), and some are both (styled divs).

In aria there's the concept of a pressed state that buttons can have. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed this requires JS to toggle on and off currently.

## Use cases

See https://github.com/openui/open-ui/issues/1039

Basically any button where a single state (which should be reflected by a fixed label) is being toggled. e.g. a mute button.

### New `"press"` value for the button element's type attribute.

Functionality this would provide atop the "button" value:

- Implicit pressed state, defaulting to false. This corresponds to aria-pressed attribute.

- Pressing the button would toggle the pressed state (e.g. false -> true for first click)

- The buttons default styling would include an indicator of this pressed state.

###  New `:pressed` pseudo class

This would apply to press buttons which are in the pressed state.
This would be used by the UA stylesheet to provide the aforementioned styling indicator.

When in the mixed state this element would match `:indeterminate` and *not* pressed.

### New `defaultpressed` DOM attribute

This would be an enum attribute and would allow the default to be aria-pressed="true" or aria-pressed="mixed" rather than "false".

It would be reflected to a new defaultPressed IDL attribute.

Ergonomically it might be good if the missing value default was "true"?

### New `pressed` IDL property

This would be a read/write IDL property, with no corresponding DOM attribute. It would get/set the internal pressed state of the button. Accepted values would be "true", "false", "mixed".

We could alternatively make pressed be boolean and use a separate indeterminate property (like checkbox) but I've heard this might not be desirable DX.

### Reuse checkbox

HTML already has a sort of state toggle, the checkbox input. For the switch component, the HTML proposal (And recently shipped implementation in Safari) uses a new attribute on the checkbox input.
Why don't we follow the same path here?

1. A switch and a checkbox are generally intended to be form participants, press buttons are *NOT* (as such my proposal introduces no form value concept for press buttons)

2. Press buttons are also *buttons* for this reason it doesn't make sense to use input as the base element.

4. (opinion) The switch should probably have been a new input type, a button type/attribute addition or a new element altogether. The selling point is that it degrades gracefully to a checkbox but a checkbox and switch have different semantics (delayed action vs immediate action)

### New press attribute on `<button type="button">`

Why am I proposing a new type rather than a new attribute for buttons?

One downside to a new button type is that invalid button types resolve to submit *not* button, this means in older browsers these press buttons could erroneously submit forms.

However, as previously stated press buttons aren't form participants and so might not be used inside of a form element as much. They also (currently) still require some JS to do an action and so you can prevent default the submission in older browsers.

Aside from the potential form submission in unsupporting browsers the new type feels like an ergonomic win compared to a new attribute.

If we take a default pressed button as an example:

`<button type="press" defaultpressed="true">` vs `<button type="button" press defaultpressed>`

Using a new type also means that we can't try and turn reset or submit or _future_ button into a toggle button accidentally.

### New pressed enum attribute?

This is the strongest alternative imo. It would alleviate the requirement to have a defaultpressed attribute and IDL property.

`<button type="button" pressed="true">` where pressed takes "true", "false", and "mixed", this would reflect the IDL property (and change when the button is pressed too).

While it's potentially clearer, it does rely on state reflection, which is undesirable imo, it also doesn't match most other elements on the web (aside from dialog and details).

## Questions

Should `:pressed` also apply if an explicit aria-pressed attribute is applied to elements with a role button?
- I'm not aware of any other state pseudo doing this.

Should this fire an event? If so, which?

How should defaultpressed attribute work if you change it once parsed? Should pressed update to match until the user clicks the button? (this is I believe how checkbox works)

Should defaultpressed instead be called initiallypressed? This might depend on the previous question. default is more consistent?

How does this work with invokers?

What should the default styling be?

What should the IDL attributes do for non-toggle buttons?

If the button is mixed what should clicking it do?

<details>
<summary>Old Version</summary>

# Toggle Buttons

Firstly I'd like to acknowledge that the term "toggle button" is incredibly overloaded, so the TLDR is a native "aria-pressed" toggle button.

## Problem

It's common in UI design to have buttons which toggle a binary state.  These are implemented in varying ways across the web, some involve JS (aria-pressed, or changing the button label), some are semantically questionable (checkbox, switch), and some are both (styled divs).

In aria there's the concept of a pressed state that buttons can have. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed this requires JS to toggle on and off currently.

## Use cases

See https://github.com/openui/open-ui/issues/1039

Basically any button where a single state (which should be reflected by a fixed label) is being toggled. e.g. a mute button.

### New `"toggle"` value for the button element's type attribute.

Functionality this would provide atop the "button" value:

- Implicit pressed state, defaulting to false. This corresponds to aria-pressed attribute.

- Pressing the button would toggle the pressed state (e.g. false -> true for first click)

- The buttons default styling would include an indicator of this pressed state.

Note: I'm intentionally ignoring the "mixed" value of the aria-pressed attribute as I'm not sure what actual utility is provides and how much it's used.

###  New `:pressed` pseudo class

This would apply to toggle buttons which are in the pressed state.
This would be used by the UA stylesheet to provide the aforementioned styling indicator.

### New `defaultpressed` DOM attribute

This would be a boolean attribute and would change the default to aria-pressed="true" rather than false.

It would be reflected to a new defaultPressed boolean IDL attribute.

### New `pressed` IDL property

This would be a read/write boolean IDL property, with no corresponding DOM attribute. It would get/set the internal pressed state of the button.

### Reuse checkbox

HTML already has a sort of state toggle, the checkbox input. For the switch component, the HTML proposal (And recently shipped implementation in Safari) uses a new attribute on the checkbox input.
Why don't we follow the same path here?

1. A switch and a checkbox are generally intended to be form participants, toggle buttons are *NOT* (as such my proposal introduces no form value concept for toggle buttons)

2. Toggle buttons are also *buttons* for this reason it doesn't make sense to use input as the base element.

### New toggle attribute on `<button type="button">`

Why am I proposing a new type rather than a new attribute for buttons?

One downside to a new button type is that invalid button types resolve to submit *not* button, this means in older browsers these toggle buttons could erroneously submit forms.

However, as previously stated toggle buttons aren't form participants and so probably won't be used inside of a form element.

Aside from the potential form submission in unsupporting browsers the new type feels like an ergonomic win compared to a new attribute.

If we take a default pressed button as an example:

`<button type="toggle" defaultpressed>` vs `<button type="button" toggle defaultpressed>`

Using a new type also means that we can't try and turn reset or submit or _future_ button into a toggle button accidentally.

#### Why can't we just add a new `pressed` attribute?

The button needs to hold a default pressed state (else AT wouldn't announce it as a toggle button unless in the pressed state).

If we just had `pressed` to indicate aria-pressed="true" then this wouldn't be possible.
</details>

**Comment by lukewarlow** [comment]
See https://demo.lukewarlow.dev/toggle-button.html for a basic demo on what this could look like. The demo uses `<toggle-button>` instead of `<button type="toggle">` for ease of implementation.

**Comment by scottaohara** [comment]
closing https://github.com/openui/open-ui/issues/1039 in favor of this issue.

**Comment by lukewarlow** [comment]
Great question, yes the intention is that this can work with invokers. We need to work out the specifics. Like should invoking a popover work with a toggle button? My gut is that this maybe shouldn't work because else you'd have two states fighting each other (expanded vs pressed). On the flip side there's some actions that maybe should require a toggle button, like video mute?

Perhaps invokers could handle the pressed state separately and not require type toggle at all. All great questions that I hope to work through a bit during today's OpenUI meeting, and further ongoing discussions.

**Comment by lukewarlow** [comment]
Some questions to think through and act as discussion points:

Should `:pressed` also apply if an explicit aria-pressed attribute is applied to elements with a role button?
- I'm not aware of any other state pseudo doing this.

Should this fire an event? If so, which?

How should defaultpressed attribute work if you change it once parsed? Should pressed update to match until the user clicks the button?

How does this work with invokers?

What should the default styling be?

Is the aria-pressed="mixed" value actually useful somehow, should this proposal try to accomodate it?

What should the IDL attributes do for non-toggle buttons?

**Comment by mfreed7** [comment]
See also: https://github.com/openui/open-ui/issues/700

**Comment by lukewarlow** [comment]
One alternative that came up in discussions is a pressed attribute but rather than a Boolean attribute it's enumerated so it's true, false or mixed (if that's needed still need to work that out).

So that solves the how to signify it's a toggle when off problem mentioned above. Would mean reflecting the value back to the attribute which might not be ideal? But also means no need for the defaultpressed attribute?

**Comment by scottaohara** [comment]
The indeterminate state should be included with this. Though not frequently used, if it is ever needed and not included then people would have to completely roll their own control, since aria states are meant to be overridden by native features. 

Eg a native checkbox with aria-checked=mixed won’t actually work since the “checked” state of the input takes priority over the aria attribute

**Comment by aardrian** [comment]
This is going to recap some of what we [discussed on Masto](https://toot.wales/@Lukew/112537552963180210).

> In aria there's the concept of a pressed state that buttons can have. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed

I encourage you to link to the specs since MDN is sometimes wrong:
https://w3c.github.io/aria/#aria-pressed

> Note: I'm intentionally ignoring the "mixed" value of the aria-pressed attribute as I'm not sure what actual utility is provides and how much it's used.

If you want parity with `aria-pressed`, then you cannot ignore `mixed`. Not accounting for it at this step means you are creating a divergent control.

As for not being sure about its utility nor how much it is used, I feel that is necessary to identify _before_ writing this proposal. In short, this proposal is premature without it.

Quick use cases:
* Expand / collapse all control
* Select / unselect all control
* Airplane mode setting
* Some of the highlighted text is bold but not all

> New `:pressed` pseudo class

And `:mixed` or lean on the existing `:indeterminate`.

> New defaultpressed DOM attribute
> This would be a boolean attribute and would change the default to aria-pressed="true" rather than false.

Needs to support mixed.

> New pressed IDL property
> This would be a read/write boolean IDL property, with no corresponding DOM attribute. It would get/set the internal pressed state of the button.

Also not boolean because it also needs to supported mixed.

> For the switch component, the HTML proposal (And recently shipped implementation in Safari) uses a new attribute on the checkbox input.

Additional notes on `switch` beyond what you provided:
* It is a binary control, meaning it does not allow a mixed state ([`switch` role in ARIA](https://w3c.github.io/aria/#switch));
* it is scoped to checkboxes;
* the Safari understanding / pitch [had some accName issues](https://bugs.webkit.org/show_bug.cgi?id=270377) that we don't want to replicate here.

> However, as previously stated toggle buttons aren't form participants and so probably won't be used inside of a form element.

They will. Partly because many authors go for a control based on look, and if this _looks_ like the control they want then they will stuff it in a form (think of cookie consent forms). If it looks the same as a switch but allows a third state, then even more reason I expect it will get used in a form. As such, it will be critical to provide detailed author guidance.

> The button needs to hold a default pressed state (else AT wouldn't announce it as a toggle button unless in the pressed state).

This is an oversimplification.
1. By "AT" I assume you mean screen readers. 
2. Screen readers get their information from the browser. 
3. That information comes from how the browser maps author-defined properties / attributes / values / states to platform accessibility APIs. 
4. Screen readers also pair with Braille displays (so "announcement" is inaccurate).
5. Voice control users are also impacted by the control type when speaking it.
6. System colors will also need to be mapped (so the control type matters here too).
7. Keyboard interaction.
8. Screen reader quick navigation keys.
9. Probably more but it's Sunday.

So the button _doesn't_ need to hold a "pressed" state if you are minting a new control. Instead, you will need to work the [HTML-AAM spec](https://w3c.github.io/html-aam/) and [ARIA spec](https://w3c.github.io/aria/) to identify how that maps.

In other words, the statement "else AT wouldn't announce" is only true if you explicitly specify it _not_ to. Or you don't collaborate, of course.

> If we just had `pressed` to indicate aria-pressed="true" then this wouldn't be possible.

So far this proposal assumes a binary control (true/false). But since you are trying to replicate ARIA, you are making a **trinary** control (true/false/mixed). Which means you might want to consider `pressed="true|false|mixed"` to mimic ARIA.

**Comment by lukewarlow** [comment]
Based on that discussion I'll write an updated version of the proposal and post it as a comment here and update the original post.

To clarify when I say the pressed Boolean attribute won't work I mean assuming it's just a standard button under the hood exactly because the aam mappings won't exist for it without some new indicator.

I appreciate the feedback this is exactly the conversation I wanted to spark with this issue.

**Comment by aardrian** [comment]
> Worth keeping in mind OpenUI is an incubator group so I feel the barrier for a "proposal" here is lower […]

My mistake. I am coming at this from a standards world.

**Comment by mfreed7** [comment]
Just to collect related things:
- https://github.com/openui/open-ui/issues/700
- https://github.com/openui/open-ui/issues/957
- https://github.com/openui/open-ui/issues/1039
- https://github.com/openui/open-ui/issues/1040
- https://github.com/whatwg/html/issues/10206
- [The CSS Toggles API proposal](https://tabatkins.github.io/css-toggle) (and [why we stopped pursuing that](https://docs.google.com/document/d/10JxXgeDAJ5zerOdxoWM9ZU9d5iMOdzGJt0EhlhkOkxM/edit))

**Comment by keithamus** [comment]
https://github.com/whatwg/html/issues/10462

**Comment by mfreed7** [comment]
Thanks. I added both of those to my list above.

**Comment by ggedde** [comment]
I love this idea and greatly wish it was available now. I think having `<button type="toggle">` is a elegant and clean solution. 
I would argue that having the button checked should include the `:checked` pseudo class. I am not sure if :pressed would be needed at that point, but not sure if that would be better for aria or not.

I would also like to have the 'for' attribute on it like `<button type="toggle" for="idref">` which would also add some identifier to the `idref` that is accessible with CSS and also be present for accessibility. Lastly if there are multiple toggle buttons with the same `for="idref"` that ALL toggle buttons toggle `:checked` whenever the `idref` changes its state. 
I know this could be included with the Invoker API and could use the `command="toggle"` and 'commandfor="idref"` as well, but I prefer the clean implementation of `for="idref"` better as it infers only only command which is toggle.

**Comment by lukewarlow** [comment]
> I know this could be included with the Invoker API and could use the command="toggle" and 'commandfor="idref"as well, but I prefer the clean implementation offor="idref"` better as it infers only only command which is toggle.

Fwiw the for attribute is probably not usable from a webcompat perspective. Hence why command is commandfor not just for directly.

As for the grouping and what not that's definitely an idea to think about.

I'll be adding a proper proposal to the site based on this issue soon, and will make sure to file follow up issues on some of these discussion points.

**Comment by ggedde** [comment]
I might have missed it in the IRC log of that discussion, but one big reason for `<button>` as opposed to `<input>` is the fact we can wrap text with it or additional tags/icons, etc if needed. So I think that should really oppose the thought of using an `<input>` like checkbox etc.

However, there is mentions that buttons are not form elements and I can see this new type being used for form inputs like we have for `<input type="button">`. Therefore I am also proposing that whatever gets added to `<button>` also be available to `<input type="button">`.

So if we do accept
```html
<button type="toggle">Yes</button>
```
then we should also have
```html
<input type="toggle" value="Yes">
```
This would render the same as `<input type="button">` where the value is the text inside the button.
<br>
Or if we go with something like:
```html
<button type="button" pressed="true">Yes</button>
```
then we should also have
```html
<input type="button" value="Yes" pressed="true">
```
<br>
I think it would just be a missed opportunity if we didn't include it for forms as well.
